libraries/standard/http/include/http_client.h

```diff
@@ -0,0 +1,339 @@
+ * @return The number of bytes written or a negative network error code.
+ */
+typedef int32_t (* HTTPTransportSend_t )( HTTPNetworkContext_t context,
+                                          const void * pBuffer, 
+                                          size_t bytesToWrite );
+
+/**
+ * @brief Transport interface for reading data on the network.
+ * 
+ * This function will read up to bytesToRead amount of data from the network.
+ * If this function returns a value less than zero, then #HTTPClient_Send will
+ * return #HTTP_NETWORK_ERROR. If this function returns less than the 
+ * bytesToRead and greater than zero, then this function will be invoked again
+ * if the data in pBuffer contains a partial HTTP response message.
+ * 
+ * @param[in] context User defined context.
+ * @param[in] pBuffer Buffer to read network data into.
+ * @param[in] bytesToRead Number of bytes requested from the network.
+ * @return The number of bytes read or a negative error code.
+ */
+typedef int32_t (* HTTPTransportRecv_t )( HTTPNetworkContext_t context,
+                                          const void * pBuffer, 
+                                          size_t bytesToRead );
+
+/**
+ * @brief The HTTP Client library transport layer interface.
+ */
+typedef struct HTTPTransportInterface
+{
+    HTTPTransportRecv_t recv;
+    HTTPTransportSend_t send;
+    HTTPNetworkContext_t context;
+} HTTPTransportInterface_t;
+
+/**
+ * @brief The HTTP Client library return status.
+ */
+typedef enum HTTPStatus
+{
+    HTTP_SUCCESS = 0,
+    HTTP_INVALID_PARAMETER,
+    HTTP_NETWORK_ERROR,
+    HTTP_NOT_SUPPORTED,
+    HTTP_PARTIAL_RESPONSE,
+    HTTP_INSUFFICIENT_MEMORY,
+    HTTP_INTERNAL_ERROR,
+    HTTP_SECURITY_ALERT_TOO_MANY_HEADERS,
+    HTTP_SECURITY_ALERT_PARSER_INVALID_CHARACTER,
+    HTTP_SECURITY_ALERT_INVALID_CONTENT_LENGTH,
+    /* TODO: Add return codes as implementation continues. */
+} HTTPStatus_t;
+
+/**
+ * @brief Request headers buffer and how many headers exist in the buffer.
+ */
+typedef struct HTTPRequestHeaders
+{
+    
+    /**
+     * @brief Buffer to hold the raw HTTP request headers.
+     * 
+     * This buffer is not available to modify until #HTTPNetworkInterface_t.recv
+     * for the response is invoked.
+     * 
+     * If this buffer is shared with the response, then this buffer cannot be
+     * modified until #HTTPClient_Send() returns.
+     */
+    uint8_t* pBuffer;
+    size_t bufferLen; /**< The length pBuffer. */
+    
+    /**
+     * @brief The actual amount of headers in the buffer, in bytes. This field
+     * is updated by the HTTP Client library functions #HTTPClient_AddHeader,
+     * and #HTTPClient_InitializeRequestHeaders.
+     */
+    size_t headersLen;
+} HTTPRequestHeaders_t;
+
+/**
+ * @brief Configurations of the initial request headers.
+ */
+typedef struct HTTPRequestInfo
+{
+    /**
+     * @brief The HTTP request method i.e. "GET", "POST", "PUT", or "HEAD". 
+     */
+    const char* method;
```

Comment:
Can this just be a numeric value (from the macros defined earlier in this file) instead of a string?

Reply:
Then we close the door on a custom method

Reply:
Custom method?

Reply:
"CONNECT", "PATCH", etc, whatever their server needs.